Drupal7 base theme (node-sass)

pabloguerino opened the pull request:
Everything works except `node-sass-asset-functions` (works outside docker) but this is something really simple, so maybe it's better to write our own as it won't take long and the logic isn't hard. This is required for the ruby/compass migration.

Will continue on this tomorrow, creating PR anyway as `node-sass-asset-functions` is not always required anyway, so the container can be used and the rest can be reviewed.

Drintios commented:
I'm getting issues trying to use this container on npm update i get:

```
+-- UNMET PEER DEPENDENCY webpack@>=1.3.0 <3

```
and also 

```
npm ERR! Linux 4.4.0-59-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "update"
npm ERR! node v7.5.0
npm ERR! npm  v4.1.2
npm ERR! path /work/node_modules/gulp-eslint/node_modules/ansi-regex
npm ERR! code EXDEV
npm ERR! errno -18
npm ERR! syscall rename

npm ERR! EXDEV: cross-device link not permitted, rename '/work/node_modules/gulp-eslint/node_modules/ansi-regex' -> '/work/node_modules/ansi-regex'
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /work/npm-debug.log
```

When trying to run gulp file i get 

```
/usr/bin/docker-entrypoint.sh: line 38: node-sass-glob: not found
module.js:472
    throw err;
    ^

Error: Cannot find module 'yargs'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/work/gulpfile.js:16:23)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
```

pabloguerino commented:
### What has been fixed

fixed issues described in the PR (for yargs and sass-glob-node)

now checking up the issue with node-sass-asset-functions seems to me like it's a dependency issue.

yep, confirmed, dependency issue, now fixed.

added "&& apk del --purge make g++ python git" to keep the container small

so with this, container should be 100% finished, theme as well (maybe missing browsersync from Rafa, but not needed for project migration at least, migration can continue in the meanwhile)

-------------

### Testing

In order to test this changes, you can build up the container locally:

```
docker build -t anything/d7 d7 --no-cache
```

and run it inside the theme (d7-base-theme) it should compile without any errors

```
docker run --rm -it -v $(pwd):/work anything/d7
```

andypost commented on d7/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+#FROM skilldlabs/frontend:latest
+FROM mhart/alpine-node:7
+
+MAINTAINER Andy Postnikov
```
please remove, docker deprecated it

andypost commented on d7/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+WORKDIR /work
+
+RUN apk add --no-cache make g++ git python && \
+    apk --no-cache --repository http://dl-cdn.alpinelinux.org/alpine/edge/community add libsass && \
```
would be great to move `add` right after `apk`

andypost commented on d7/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+COPY docker-entrypoint.sh /usr/bin/
+ENTRYPOINT ["docker-entrypoint.sh"]
+CMD ["gulp"]
+
```
no reason in extra line)

andypost commented on d7/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+        node-sass-asset-functions && \
+     apk del --purge make g++ python git
+
+COPY docker-entrypoint.sh /usr/bin/
```
better put it in `/usr/local/bin`

pabloguerino commented:
Updated! Thanks Andy for the review

jbguerraz commented:
LGTM

andypost commented on d7/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+
+COPY docker-entrypoint.sh /usr/local/bin/
+ENTRYPOINT ["docker-entrypoint.sh"]
+CMD ["gulp"]
```
somehow there's no CR/LF now

pabloguerino commented:
Fixed crlf issue, tested container against a real project (the one that we tested the migration first) and it's all ok

iberdinsky-skilld commented:
getting errors in installation
`npm WARN deprecated minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated graceful-fs@1.2.3: graceful-fs v3.0.0 and before will fail on node releases >= v7.0. Please update to graceful-fs@^4.0.0 as soon as possible. Use 'npm ls graceful-fs' to find it in the tree.`
and
`npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/webpack/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.1: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/webpack-dev-server/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.1: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN webpack-dev-server@2.4.1 requires a peer of webpack@^2.2.0 but none was installed.`

Drintios commented:
I think first one is related to second one and the second one is caused because fs event is required if you are using Mac but in case you don't, as it not supported on other platforms, it throws error, but nothing to worry about, I think this is a dependency for gulp, but as I wrote they include this in case you will use mac

pabloguerino commented:
what Rafa said is correct, there is nothing to worry about that warning, just a message (not error) because of a dependency between different platforms that is not needed here and it's skipped

andypost commented:
It needs follow-up for local=>internal user mapping
Latest example https://github.com/tkyryliuk/drupal-boilerplate/blob/master/docker/web/entrypoint.sh#L8
